features/topology-map/org.opennms.features.topology.app/src/main/java/org/opennms/features/topology/app/internal/jung/HierarchyLayoutAlgorithm.java

```diff
@@ -0,0 +1,129 @@
+ * <p>
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ * <p>
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ * http://www.gnu.org/licenses/
+ * <p>
+ * For more information contact:
+ * OpenNMS(R) Licensing <license@opennms.org>
+ * http://www.opennms.org/
+ * http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.topology.app.internal.jung;
+
+
+import java.awt.geom.Point2D;
+import java.util.Collection;
+
+import edu.uci.ics.jung.algorithms.layout.TreeLayout;
+import edu.uci.ics.jung.algorithms.shortestpath.MinimumSpanningForest;
+import edu.uci.ics.jung.graph.DelegateForest;
+import edu.uci.ics.jung.graph.Forest;
+import edu.uci.ics.jung.graph.SparseGraph;
+import org.opennms.features.topology.api.Graph;
+import org.opennms.features.topology.api.GraphContainer;
+import org.opennms.features.topology.api.Layout;
+import org.opennms.features.topology.api.Point;
+import org.opennms.features.topology.api.topo.Edge;
+import org.opennms.features.topology.api.topo.Vertex;
+import org.opennms.features.topology.api.topo.VertexRef;
+
+/**
+ * Algorithm to transform a graph of vertices and edges into a hierarchical structure of a 2D plane.
+ * It is a geometric transformation which does not change the graph but the x an y position of its vertices.
+ *
+ * The current graph to be transformed is extracted from the {@link org.opennms.features.topology.api.GraphContainer}.
+ * Basically all its vertices are transformed and then the new layout is set back to the GraphContainer by executing
+ * the {@link #updateLayout(GraphContainer) updateLayout} method.
+ *
+ *
+ * <p>
+ *     Currently these algorithms can be used for transformations:
+ *     <ul>
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.CircleLayoutAlgorithm}
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.D3TopoLayoutAlgorithm}
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.FRLayoutAlgorithm}
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.ISOMLayoutAlgorithm}
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.KKLayoutAlgorithm}
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.RealUltimateLayoutAlgorithm}
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.SpringLayoutAlgorithm}
+ *          <li>{@link org.opennms.features.topology.app.internal.jung.TopoFRLayoutAlgorithm}
+ *     </ul><p>
+ *
+ * @see <a href="http://jung.sourceforge.net">Java Universal Network/Graph Framework</a>
+ *
+ *
+ */
+public class HierarchyLayoutAlgorithm extends AbstractLayoutAlgorithm {
```

Comment:
Please add a javadoc statement about the assumptions about the provided graph:
* The algorithm generates the layout correctly if only one root element exists or if there are no vertices at all
* If no root element exists, a random vertex from the graph is picked and defined as root.
* If multiple root elements exist, the first one returned by `getDisplayedVertices()` which has no parent (= root candidate) is picked as the root element